src/main/java/uk/ac/cam/cl/dtg/teaching/pottery/model/Step.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * pottery-backend-interface - Backend API for testing programming exercises
+ * Copyright © 2015-2018 BlueOptima Limited, Andrew Rice (acr31@cam.ac.uk)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package uk.ac.cam.cl.dtg.teaching.pottery.model;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.wordnik.swagger.annotations.ApiModelProperty;
+
+import java.util.Map;
+
+/**
+ * A Step represents a step that Pottery carries out when processing the user's code.
+ *
+ * A Step has a name which is used to identify it's output.
+ *
+ * A Step has a map of variant names to Executions for that variant. If no Execution is provided for
+ * a variant, then the Execution for the "default" variant is run, if it exists. If it does not
+ * exist, no Execution is run for this step.
+ *
+ * For example, a compile Step might only have Executions specified for variants that require
+ * compilation, and a validation Step might only have a default variant so that all variants are
+ * validated with the same code.
+ */
+public class Step {
+  @ApiModelProperty("Name of this step.")
+  private String name;
+  @ApiModelProperty("Map from variants to execution to run for this step.")
+  private Map<String, Execution> execution;
+
+  @JsonCreator
+  public Step(
+      @JsonProperty("name") String name,
+      @JsonProperty("execution") Map<String, Execution> execution) {
+    this.name = name;
+    this.execution = execution;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public Map<String, Execution> getExecution() {
```

Comment:
getExecutionMap?

Reply:
it would be nice to be able to return ImmutableMap here - does jackson get it right if we set the member variable type to ImmutableMap?

Reply:
After a bit of faffing, I found it doesn't work unless you add in https://github.com/FasterXML/jackson-datatypes-collections/tree/master/guava but then you need to upgrade to a compatible Jackson version. At that point I figured this was more effort than it was worth. What do you think?

Reply:
I agree - more effort than its worth.